src/components/calcite-dropdown/calcite-dropdown.tsx

```diff
@@ -77,20 +81,52 @@ export class CalciteDropdown {
       "[slot=dropdown-trigger]"
     ) as HTMLSlotElement;
     if (!this.sorted) {
-      this.items = this.sortItems(this.items);
+      const groups = this.items.sort(
+        (a, b) => a.position - b.position
+      ) as GroupRegistration[];
+
+      const { maxItemsToDisplay } = this;
+      let itemsToProcess = 0;
+
+      groups.forEach((group) => {
+        if (maxItemsToDisplay > 0 && itemsToProcess < maxItemsToDisplay) {
+          this.maxScrollerHeight += group?.titleEl?.offsetHeight || 0;
+
+          group.items.forEach((item) => {
+            if (itemsToProcess < maxItemsToDisplay) {
+              this.maxScrollerHeight += item.offsetHeight;
+              itemsToProcess += 1;
+            }
+          });
+        }
+      });
+
+      this.items = groups.reduce(
+        (items, group) => [...items, ...group.items],
+        []
+      );
```

Comment:
Nitpick - I'm not sure if this is valid use case currently, but it might be nice to pull this out into its own function we can run on `didUpdate` or something, if the content of an item changes or an icon is added via prop it may cause the item to wrap and the maxHeight needs to be re-calculated. 

If we address the open PR to allow the width of the dropdown to match the width of the trigger, which itself may change, this could be needed too.

Reply:
I extracted the method, but the component will still need some refactoring to keep a reference to the registered groups. We can tackle that once the use case you mentioned comes into play. 💪